### PR TITLE
feat(atenlib): test random ops

### PR DESCRIPTION
**Pull request by justinchuby**
Check shape and dtype only for ops that are known to be indeterministic.
Also added the skip_if mechanism to skip tests based on environment.

### CODE REVIEW

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 1]
```diff
@@ -402,25 +430,17 @@ def _where_input_wrangler(
...
-                "nn.functional.upsample_nearest2d",
-                reason="ONNX Runtime 1.13 does support opset18",
-            ),
-        )
-        if version_utils.onnxruntime_older_than("1.14")
-        else ()
+    xfail(
+        "logsumexp",
+        reason="ONNX Runtime 1.13 does not support ReduceLogSumExp-18",
+        active_if=version_utils.onnxruntime_older_than("1.14"),
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:436]: Do we need this or should we just remove the xfail?

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 2]
```diff
@@ -389,6 +406,17 @@ def _where_input_wrangler(
     "zeros_like": core_ops.aten_zeros_like,
 }
 
+# These ops are not deterministic, so we check shape and dtype only
+INDETERMINISTIC_OPS: frozenset[str] = frozenset(
```
Comment (gramalingam) [inline onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:410]: Nit: "Indeterministic" => "Nondeterministic"

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:410]: Thanks 😅

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:436]: For now we still need it because the ort version with function default attribute support is still a variant of 1.13. @gramalingam i think we can move to use nightly after liqun s PR?

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py]: Done

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 3]
```diff
@@ -743,14 +765,27 @@ def test_output_match(self, device: str, dtype: torch.dtype, op):
...
+                    expected = (
+                        torch_output
+                        if isinstance(torch_output, torch.Tensor)
+                        else torch.tensor(torch_output)
+                    )
+
+                    if op.name in NONDETERMINISTIC_OPS:
+                        # Check shape and dtype only for ops that are known to be
+                        # nondeterministic
+                        self.assertEqual(actual.shape, expected.shape)
```
Comment (fatcat-z) [inline onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:778]: It'd better to verify the mean value between them are very close with an acceptable tolerance. [Here](https://github.com/pytorch/pytorch/blob/a641d6075761050599bf972c0b6bc3823182914e/test/onnx/test_pytorch_onnx_onnxruntime.py#L11792)is an example.

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:778]: Ops like empty are creating non initialized values. Or dropout which does not expect a matching mean. Should we distinguish them?

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:778]: I think I am inclined to create individual tests for distributions if desired

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 4]
```diff
@@ -97,16 +101,18 @@ def skip(
...
 
     Args:
         op_name: The name of the operator.
         variant_name: Optional OpInfo variant_test_name.
-        dtypes: The dtypes to skip.
         reason: The reason for skipping.
+        dtypes: The dtypes to skip.
         matcher: A function that matches the test sample input. It is used only when
             xfail is in the SKIP_SUBTESTS list.
+        active_if: Whether the skip is active.
```
Comment (fatcat-z) [inline onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:115]: A little confused: I see this parameter was used in xfail() function, while this description is saying about skip.

Is this a general parameter, or related to skip or xfail?

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:115]: Sorry that was a typo. It’s general and inherited from PyTorch

**Comment by justinchuby** [comment]
I removed merge at lgtm because there seems to be more comments to address

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:115]: Updated. I didn't create the name but it makes sense: when used it becomes something like `active_if=onnxruntime_older_than("1.14"),`, which is good for redability.

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 3 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:778]: Created https://github.com/microsoft/onnx-script/issues/467 to track

**Comment by justinchuby** [comment]
I think I addressed everything.

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 4 above)
Comment (fatcat-z) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:115]: Nit:
I agree that "active_if" also makes sense. 

I'm thinking this in terms of unified style of our code. While I go through our code base, we didn't use "active" to describe if something will work or not. Most of time, we will use "enabled" for such purpose, like cudnn_enabled, logger.isEnabledFor. 

So probably "enabled_if" would be better.

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:115]: SG!

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py]: Done

**Review by fatcat-z — APPROVED** [review]
LGTM, thanks!
